### PR TITLE
Feat/brave search

### PR DESCRIPTION
**Pull request by ManagarmrWhisperer**
Some cleaning and testing should be done, especially for hiding the results. The logic for running should be improved, since the page gets re-rendered two times it shouldn't be necessary to use a MO.

- [ ] Improve running logic
- [x] Cleaning up the code
- [x] Checking result hiding.

### CODE REVIEW

File: js/baseSearch.js [hunk 1]
```diff
@@ -152,3 +152,14 @@ export function awaitElement( knownParent, selector, callback ) {
         childList: true
     } );
 }
+
+export function observeElement( selector, config, callback ) {
+    // This isn't a good idea
+    config = config || { attributes: false, childList: true, subtree: true };
```
Comment (alex4401) [inline js/baseSearch.js:158]: I don't see much use in this helper other than possibly lowering efficiency.

File: js/braveSearch.js [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        element.setAttribute( this.MARKER_ATTRIBUTE, '1' );
+    },
+
+
+    isLocked( element ) {
+        return element.getAttribute( this.MARKER_ATTRIBUTE ) === '1';
+    },
+
+
+    makePlaceholderElement( wiki ) {
```
Comment (alex4401) [inline js/braveSearch.js:52]: Should be extracted into a common module.

File: js/braveSearch.js [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        }
+    }
+};
+
+
+// Rewrites a Fandom result to an official wiki link to help users switch
+
+const rewrite = {
+    MARKER_ATTRIBUTE: 'data-lock',
+    ENGINE_LAYOUT_SELECTOR: '#results',
```
Comment (alex4401) [inline js/braveSearch.js:104]: These selectors should be moved into constants if shared with filter.

File: js/braveSearch.js [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    URL_ELEMENT_SELECTOR: '.snippet-url',
+    SPAN_TITLE_ELEMENT_SELECTOR: '.heading-serpresult',
+    BADGE_ELEMENT_SELECTOR: '.svelte-1ckzfks', // Which element shall contain the Badge?
+    ANCHOR_ELEMENT_SELECTOR: 'a.svelte-1dihpoi',
+
+
+    makeBadgeElement( isTopLevel ) {
+        const out = document.createElement( 'span' );
+        out.innerText = isTopLevel ? 'redirected' : 'some redirected';
+        out.style.backgroundColor = document.querySelector( 'html.dark' )
```
Comment (alex4401) [inline js/braveSearch.js:115]: document.documentElement.classList.has( 'dark' )

File: js/braveSearch.js [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        out.style.textDecoration = 'none';
+        out.style.verticalAlign = 'middle';
+        out.classList.add( 'rewrite_badge' );
+        return out;
+    },
+
+
+    rewriteLink( wiki, link ) {
+        if ( link.href.startsWith( '/url?' ) ) {
+            link.href = ( new URLSearchParams( link.href ) ).get( 'url' );
```
Comment (alex4401) [inline js/braveSearch.js:133]: Are you sure this is what you want done here?

File: js/braveSearch.js [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+                this.rewriteTitle( wiki, child );
+            }
+        }
+    },
+
+
+    rewriteURLElement( wiki, node ) {
+        for ( const child of node.childNodes ) {
+	    // Checks if the child is a protocol, some engines split the url in: <proto://><domain>
+            if ( /(?<=.+):\/\//.test( child.textContent ) ) {
```
Comment (alex4401) [inline js/braveSearch.js:159]: Is this actually needed? What is it trying to solve exactly?

File: js/braveSearch.js [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+                continue;
+            }
+            child.textContent = child.textContent.replace( new RegExp( `${wiki.oldId || wiki.id}.(fandom|gamepedia)?.com` ), `${wiki.id}.wiki.gg` );
```
Comment (alex4401) [inline js/braveSearch.js:162]: Make the regexp a local constant.

File: js/braveSearch.js [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+                this.rewriteTitle( wiki, span );
+            }
+
+            // Rewrite URL element
+            for ( const url of element.querySelectorAll( this.URL_ELEMENT_SELECTOR ) ) {
+                this.rewriteURLElement( wiki, url );
+            }
+
+            element.prepend( this.makeBadgeElement( isTopLevel ) );
```
Comment (alex4401) [inline js/braveSearch.js:202]: You're passing a closure to the function, and not a bool. This is not doing what you want it doing.

I don't see any reasons for the closure either.

File: js/braveSearch.js [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+            if ( child.textContent && child.textContent.length >= 5 ) {
+                child.textContent = this.rewriteText( wiki, child.textContent );
+            } else {
+                this.rewriteTitle( wiki, child );
+            }
+        }
+    },
+
+
+    rewriteURLElement( wiki, node ) {
```
Comment (alex4401) [inline js/braveSearch.js:156]: `rewriteUrlElement`

File: manifest_chrome.json
Comment (alex4401) [inline manifest_chrome.json:1]: This file was wrongly moved into repository root.

File: js/braveSearch.js (hunk 2 above)
Comment (ManagarmrWhisperer) [reply js/braveSearch.js:52]: Should we move  `makeBadgeElement` too?

File: js/braveSearch.js (hunk 5 above)
Comment (ManagarmrWhisperer) [reply js/braveSearch.js:133]: Yep, we need that check!

**Review by alex4401 — CHANGES_REQUESTED** [review]
Left first review comments.

File: js/braveSearch.js (hunk 2 above)
Comment (alex4401) [reply js/braveSearch.js:52]: Yes.

File: js/braveSearch.js [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+
+    isLocked( element ) {
+        return element.getAttribute( this.MARKER_ATTRIBUTE ) === '1';
+    },
+
+
+
+    run( wiki, linkElement ) {
+        if ( linkElement.parentElement && document.querySelector( this.ENGINE_LAYOUT_SELECTOR ).contains( linkElement ) ) {
```
Comment (alex4401) [inline js/braveSearch.js:55]: Write a comment what this is for.

File: js/braveSearch.js [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+                    resultContainer.insertBefore( officialResult, oldElement );
+                }
+            } else {
+                // Insert a placeholder before this result
+                const resultContainer = document.querySelector( this.ENGINE_LAYOUT_SELECTOR );
+                resultContainer.insertBefore( makePlaceholderElement( wiki ), oldElement );
+            }
+
+            // Hides the main result element
+            oldElement.style.display = 'none';
```
Comment (alex4401) [inline js/braveSearch.js:86]: Other modules remove the result container. Why hide here? This should be explained in the source.

File: js/braveSearch.js [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+            oldElement.parentElement.prepend( makePlaceholderElement( wiki ), oldElement );
```
Comment (alex4401) [inline js/braveSearch.js:87]: This will spawn placeholders at the top of the results container, won't it?

File: js/braveSearch.js [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+                const officialResult = findNextOfficialWikiResult( wiki, oldElement, this.ENGINE_RESULT_CONTAINER_SELECTOR );
+                if ( officialResult ) {
+                    // Move the official result before this one
+                    const resultContainer = document.querySelector( this.ENGINE_LAYOUT_SELECTOR );
+                    resultContainer.insertBefore( officialResult, oldElement );
+                }
+            } else {
+                // Insert a placeholder before this result
+                const resultContainer = document.querySelector( this.ENGINE_LAYOUT_SELECTOR );
+                resultContainer.insertBefore( makePlaceholderElement( wiki ), oldElement );
```
Comment (alex4401) [inline js/braveSearch.js:82]: Can't you access the results container via parentNode if oldElement is its child anyway?

File: js/braveSearch.js [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+            if ( topLevelLinkElement && !topLevelLinkElement.href.startsWith( `https://${wiki.oldId || wiki.id}.fandom.com` ) ) {
+                return;
+            }
+
+            if ( oldElement !== null ) {
+                // Find an official wiki result after this one
+                const officialResult = findNextOfficialWikiResult( wiki, oldElement, this.ENGINE_RESULT_CONTAINER_SELECTOR );
+                if ( officialResult ) {
+                    // Move the official result before this one
+                    const resultContainer = document.querySelector( this.ENGINE_LAYOUT_SELECTOR );
```
Comment (alex4401) [inline js/braveSearch.js:76]: You've already run this query before.

File: js/braveSearch.js [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+            this.lock( linkElement );
+        }
+    }
+};
+
+function runCallback() {
+    invokeSearchModule( wikis, rewrite.run.bind( rewrite ), filter.run.bind( filter ) );
+
+}
+observeElement( '#results', { attributes: false, childList: true, subtree: true }, runCallback );
```
Comment (alex4401) [inline js/braveSearch.js:177]: A closure will be sufficient here, and cleaner.

File: manifests/chrome_dev.json [hunk 16]
```diff
@@ -74,7 +74,17 @@
...
+	{
+            "matches": [
+                "https://search.brave.com/*"
+            ],
+            "js": [
+                "built/braveSearch.js"
+            ],
+            "run_at": "document_end"
         }
+
```
Comment (alex4401) [inline manifests/chrome_dev.json:87]: Redundant new line.

**Review by alex4401 — CHANGES_REQUESTED** [review]

File: js/baseSearch.js [hunk 17]
```diff
@@ -152,3 +152,37 @@ export function awaitElement( knownParent, selector, callback ) {
...
+    element.style.color = '#5f6368';
+    element.classList.add( 'filter_badge' );
+    element.style.padding = '0px 0px 1em 10px';
+    element.style.display = 'block';
+    return element;
+}
+export function makeBadgeElement( isTopLevel ) {
+        const out = document.createElement( 'span' );
+        out.innerText = isTopLevel ? 'redirected' : 'some redirected';
+        out.style.backgroundColor = document.documentElement.classList.has( 'dark' )
```
Comment (alex4401) [inline js/baseSearch.js:167]: This is not generic.

File: js/baseSearch.js [hunk 18]
```diff
@@ -152,3 +152,37 @@ export function awaitElement( knownParent, selector, callback ) {
...
+export function makePlaceholderElement( wiki ) {
+    const element = document.createElement( 'span' );
+    element.innerHTML = 'Result from ' + wiki.search.placeholderTitle + ' hidden by wiki.gg redirector';
+    element.style.color = '#5f6368';
+    element.classList.add( 'filter_badge' );
+    element.style.padding = '0px 0px 1em 10px';
+    element.style.display = 'block';
+    return element;
+}
+export function makeBadgeElement( isTopLevel ) {
```
Comment (alex4401) [inline js/baseSearch.js:164]: Spacing.

File: js/baseSearch.js [hunk 19]
```diff
@@ -152,3 +152,37 @@ export function awaitElement( knownParent, selector, callback ) {
         childList: true
     } );
 }
+export function makePlaceholderElement( wiki ) {
```
Comment (alex4401) [inline js/baseSearch.js:155]: Spacing

File: js/baseSearch.js [hunk 20]
```diff
@@ -152,3 +152,37 @@ export function awaitElement( knownParent, selector, callback ) {
...
+        out.style.fontSize = '70%';
+        out.style.borderRadius = '4px';
+        out.style.padding = '1px 6px';
+        out.style.marginLeft = '0px';
+        out.style.opacity = '0.6';
+        out.style.textDecoration = 'none';
+        out.style.verticalAlign = 'middle';
+        out.classList.add( 'rewrite_badge' );
+        return out;
+    }
```
Comment (alex4401) [inline js/baseSearch.js:180]: Indents.

File: js/baseSearch.js [hunk 21]
```diff
@@ -152,3 +152,37 @@ export function awaitElement( knownParent, selector, callback ) {
...
+
+export function observeElement( selector, config, callback ) {
+    const targetElement = document.querySelector( selector );
+    const observer = new MutationObserver( callback );
+
+    observer.observe( targetElement, config );
+
```
Comment (alex4401) [inline js/baseSearch.js:187]: Redundant new line. This method should also return the observer.

File: js/braveSearch.js [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+
+            // Rewrite anchor href links
+            for ( const a of element.getElementsByTagName( 'a' ) ) {
+                this.rewriteLink( wiki, a );
+            }
+	    
+            // Rewrite title and append a badge
+            for ( const span of element.querySelectorAll( this.SPAN_TITLE_ELEMENT_SELECTOR ) ) {
+                if ( !wiki.search.titlePattern.test( span.textContent ) ) {
+                    continue;
```
Comment (alex4401) [inline js/braveSearch.js:156]: Invert the if. The rewriteTitle call can happen if the condition is met, not if the negated condition is not met.

File: rollup-config-background.mjs [hunk 23]
```diff
@@ -6,6 +6,8 @@ const scripts = [
     { script: 'popup' },
     { script: 'search/google', output: 'google', isContentScript: true },
     { script: 'search/ddg', output: 'ddg', isContentScript: true },
+    { script: 'search/braveSearch', output: 'braveSearch', isContentScript: true },
+    { script: 'search/ddg', output: 'ddg', isContentScript: true },
```
Comment (alex4401) [inline rollup-config-background.mjs:10]: The DDG module got duplicated here.

File: js/search/braveSearch.js [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+            for ( const a of element.getElementsByTagName( 'a' ) ) {
+                this.rewriteLink( wiki, a );
+            }
+
+            // Rewrite title and append a badge
+            for ( const span of element.querySelectorAll( this.SPAN_TITLE_ELEMENT_SELECTOR ) ) {
+                if ( wiki.search.titlePattern.test( span.textContent ) ) {
+                    this.rewriteTitle( wiki, span );
+                }
+
```
Comment (alex4401) [inline js/search/braveSearch.js:152]: Redundant new line.

**Review by alex4401 — CHANGES_REQUESTED** [review]
Rebase on top of feat/search2 and resolve remaining change comments.

**Pull request closed without merging**